pkg/metadata/main.go

```diff
@@ -154,55 +154,44 @@ func processUserData(data []byte) error {
 		// This is not an error
 		return nil
 	}
-	cm, ok := fd.(map[string]interface{})
-	if !ok {
-		log.Printf("Could convert JSON to desired format: %s", fd)
-		return nil
-	}
-	for d, val := range cm {
-		dir := path.Join(ConfigPath, d)
-		if err := os.MkdirAll(dir, 0755); err != nil {
-			log.Printf("Failed to create %s: %s", dir, err)
-			continue
-		}
-		files, ok := val.(map[string]interface{})
-		if !ok {
-			log.Printf("Could convert JSON for files: %s", val)
-			continue
-		}
-		for f, i := range files {
-			p := uint64(0644)
-			var c string
-
-			switch fi := i.(type) {
-			case map[string]interface{}:
-				if _, ok := fi["perm"]; !ok {
-					log.Printf("No permission provided %s", f)
-					continue
-				}
-				if _, ok := fi["content"]; !ok {
-					log.Printf("No content provided %s", f)
-					continue
-				}
-				c = fi["content"].(string)
-				if p, err = strconv.ParseUint(fi["perm"].(string), 8, 32); err != nil {
-					log.Printf("Failed to parse permission %s: %s", fi, err)
-					continue
-				}
-			case string:
-				c = fi
-			default:
-				log.Printf("Couldn't convert JSON for items: %s", i)
-				continue
-			}
 
-			if err := ioutil.WriteFile(path.Join(dir, f), []byte(c), os.FileMode(p)); err != nil {
-				log.Printf("Failed to write %s/%s: %s", dir, f, err)
-				continue
+	writeConfigFiles(basePath, fd)
+	return nil
+}
 
+func writeConfigFiles(dir string, input interface{}) {
+	switch json := input.(type) {
+	case map[string]interface{}:
+		if isFile(json) {
+			perm, err := strconv.ParseUint(json["perm"].(string), 8, 32)
+			if err != nil {
+				log.Printf("Failed to parse permission %s: %s", json, err)
+				return
+			}
+			if err := ioutil.WriteFile(dir, []byte(json["content"].(string)), os.FileMode(perm)); err != nil {
+				log.Printf("Failed to write %s: %s", dir, err)
+				return
+			}
+		} else {
+			if err := os.MkdirAll(dir, 0755); err != nil {
+				log.Printf("Failed to create %s: %s", dir, err)
+				return
+			}
+			for subpath, subtree := range json {
+				writeConfigFiles(path.Join(dir, subpath), subtree)
 			}
 		}
+	case string:
+		if err := ioutil.WriteFile(dir, []byte(json), os.FileMode(0644)); err != nil {
+			log.Printf("Failed to write %s: %s", dir, err)
+		}
+	default:
+		log.Printf("Couldn't convert JSON for items: %s", input)
 	}
+}
 
-	return nil
+func isFile(json map[string]interface{}) bool {
+	_, perm := json["perm"]
+	_, content := json["content"]
+	return perm && content
```

Comment:
Does this mean I cannot create a directory simultaneously containing two files named `perm` and `content`?

It would be good to see `docs/metadata.md` updated to reflect the new structure/behaviour.

Reply:
Yes that's not good. Should we enforce type of objects by adding a new field ? Like directory or file. It will be better in the future if we want to add directory permissions.

```json
{
  "type": "directory",
  "name": "etc",
  "content": [
    {
      "type": "directory",
      "name": "ssl",
      "content": [
        {
          "type": "directory",
          "name": "certs",
          "content": [
            {
              "type": "file",
              "name": "ca-certificates.crt",
              "content": "some content"
            }
          ]
        }
      ]
    },
    {
      "type": "file",
      "name": "resolv.conf",
      "content": "some content"
    }
  ]
}
```

It's a breaking change now, but maybe better.

Reply:
I'd suggest either `content` becomes bimodal (so contains either a JSON string and is a file or contains a JSON map and is a directory, which can be detected at runtime) or add a new field `entries` which would be an map (i.e. a directory) but mutually exclusive with the existing `content` field so you can tell the cases apart that way.

I think the latter option (adding `entries`) is likely to be nicer from a schema PoV (fields which a dynamically typed are a bit of a pain).

I don't think you need the `type` entry to distinguish the cases, which I think means it can be backwards compatible.

Reply:
I will write an other PR for that, thanks.

Reply:
If we merge this  PR and then update/fix in a followup then that will be two breaking changes rather than zero. Or maybe you meant for that PR to supersede/replace this one?

Reply:
:+1:, thanks!